### PR TITLE
Make SOFTWARE attribute configurable

### LINKED ISSUE #18: Add support for customizing SOFTWARE attribute value
Currently this is hard-coded and should be specifiable either through configuration or environment variables.

### PR DESCRIPTION
**Pull request by namreg**
#### Description
Added ability to configure `SOFTWARE` attribute. For a client is configurable through command argument `software`, for a server - through environment variable`SOFTWARE`.
#### Reference issue
Fixes #18

### CODE REVIEW

**Comment by Sean-Der** [comment]
Really fantastic work @namreg, this is great!

Would you mind adding a simple test just to confirm that `software` is added to requests/responses. You don't have to check everything, but just to make sure someone doesn't break your code in the future :)

thank you!

**Comment by namreg** [comment]
@Sean-Der I've added some tests :)

**Comment by Sean-Der** [comment]
Fantastic work, the tests look great!

File: sender.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package turn
+
+import (
+	"net"
+
+	"github.com/pion/stun"
+)
+
+// Sender is responsible for building a message and sending it to the given addr.
+type Sender func(conn net.PacketConn, addr net.Addr, attrs ...stun.Setter) error
```
Comment (Sean-Der) [inline sender.go]: Oh that is really interesting! I bet that could be helpful if you want to insert a custom attr/inspect in every message. Nice idea.

File: sender.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+// DefaultBuildAndSend is a sender default implementation.
+func DefaultBuildAndSend(conn net.PacketConn, dst net.Addr, attrs ...stun.Setter) error {
```
Comment (Sean-Der) [inline sender.go]: Does this need to be public?

**Comment by Sean-Der** [comment]
Fantastic work, the tests look great! Mind making that one visibility change and I will merge :)

If you are interested we would love to have you in [Slack](https://pion.ly/slack), we talk about a lot of fun P2P/WebRTC stuff there!

File: sender.go (hunk 1 above)
Comment (namreg) [reply sender.go]: Thank you! Yeah, it will allow us to analyze in-depth every message.

File: sender.go (hunk 2 above)
Comment (namreg) [reply sender.go]: No, no needs) fixed

**Comment by namreg** [comment]
> If you are interested we would love to have you in Slack, we talk about a lot of fun P2P/WebRTC stuff there!

I am already there, thanks! :)

**Comment by Sean-Der** [comment]
Fantastic work @namreg!

Thanks for the change, and I look forward to anything in the future :)

**Comment by enobufs** [comment]
@namreg 
I am working on major changes to pion/turn Client implementation, and I noticed you have added `Sender` to its config parameter, and I am trying to understand why it was necessary... It appears to me that it has nothing to do with adding Software attribute.

See #75 

@Sean-Der

**Comment by Sean-Der** [comment]
@enobufs 

Another thing that was useful was tracing all the outbound messages. I think @namreg is logging/analyzing everything that is being sent.

I think it would be ok to cut down the amount of arguments/make the API simpler as long as we can still do that!

**Comment by enobufs** [comment]
@Sean-Der 

If that's the purpose, then I would approach in a different way...  The `Sender` is taking over the entire (original) "buildAndSend" including 'conn'!  ([the code](https://github.com/pion/turn/blob/f0df9a6e8c7e52afd24155f555f8384d6e9cb8f8/client.go#L76) I think it is too intrusive for our API.

To monitor what was sent, then I would provide a pipe of net.PacketConn to tap in what was sent.

If @namreg wanted to have complete flexibility in building STUN request (e.g. adding arbitrary attributes, etc), then with the new API, we have `client.PerformTransaction` method to use.

For now, I removed Sender option from the client config in my PR #75 during the rebase I have just did. (Server still has it but I think we should remove it later and provide another way.)

**Comment by namreg** [comment]
@enobufs The main reason to add Sender was making code mode testable because I wanted to test code I have written.

> To monitor what was sent, then I would provide a pipe of net.PacketConn 

I haven't look at your PR yet, but I think `net.PacketConn` is not a very good abstraction - it is supposed using the only udp connection. But according to the rfc (also we have a corresponding issue #1 ) client and server can talk via TCP (+ TCP over TLS).

**Comment by enobufs** [comment]
When TCP is introduced, we will likely add net.Conn for TCP connection.

**Pull request merged**